File: OptibenchOptimizer/Implementations/RemoteProblem.cs
using System.Net.Http.Headers;
using interfaces;

namespace Implementations 
{
    class RemoteProblem : IProblem
    {
        public readonly string ProblemName;
        public readonly string Uri;
        private readonly HttpClient client = new HttpClient
        {
           Timeout = TimeSpan.FromSeconds(3) // ako ne dobije odgovor u roku od 3s, http zahtjev ce se prekinuti
        };

        public RemoteProblem(string uri, string problemName) 
        {
            this.ProblemName = problemName;
            this.Uri = uri;
            client.BaseAddress = new Uri(this.Uri); //za "problem",
            client.DefaultRequestHeaders.Accept.Clear();
            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        }

        public async Task<double> GetExactSolution(string problemName)
        {
            string path = $"exact-solution/{this.ProblemName}";
            HttpResponseMessage response = await client.GetAsync(path);
            double exactSolution = double.MaxValue;

            if (response.IsSuccessStatusCode)
            {
                string retSolution = await response.Content.ReadAsStringAsync();
                

                if (double.TryParse(retSolution, out double parsedSolution))
                    exactSolution = parsedSolution;
                else
                    Console.WriteLine($"Cannot parse response '{retSolution}' to double value.");

            }
            return exactSolution; //vratice NaN ako nesto ne valja, ne postoji problem ili ne postoji tacno rjesenje

        }

        public async Task<double> GetValue(double[] x)
        {
            string path = $"problems/{this.ProblemName}?{string.Join("&", x.Select(p => $"x={p}"))}";
            HttpResponseMessage response = await client.GetAsync(path);
            double problem = double.NaN;

            
            if (response.IsSuccessStatusCode)
            {
                string retProblem = await response.Content.ReadAsStringAsync();
                

                if (double.TryParse(retProblem, out double parsedProblem))
                    problem = parsedProblem;
                else
                    Console.WriteLine($"Cannot parse response '{retProblem}' to double value.");

            }
            
            return problem; //vratice NaN ako nesto ne valja
        }

        
    }
}

File: PythonOptimizer/Implementations/RemoteProblem.py
import math
from urllib.parse import urljoin
import requests
from Abstractions.IProblem import IProblem


class RemoteProblem(IProblem):
    def __init__(self, uri, problem_name):
        self.problem_name = problem_name
        self.uri = uri
        self.client = requests.Session()
        self.client.headers.clear()
        self.client.headers.update({'Accept': 'application/json'})

    async def get_value(self, x):
        path = f"problems/{self.problem_name}?" + "&".join([f"x={p}" for p in x])
        full_url = urljoin(self.uri, path)
        response = self.client.get(full_url)
        problem = math.nan

        if response.status_code == 200:
            ret_problem = response.json()

            try:
                parsed_problem = float(ret_problem)
                problem = parsed_problem
            except ValueError:
                print(f"Cannot parse response '{ret_problem}' to double value.")
        
        return problem
    
    
    async def get_exact_solution(self, problem_name):
        path = f"exact-solution/{problem_name}"
        full_url = urljoin(self.uri, path)
        exact_solution = math.inf  # Set to max by default
       

        try:
            response = self.client.get(full_url)
            if response.status_code == 200:
                ret_solution = response.text
                try:
                    exact_solution = float(ret_solution)
                except ValueError:
                    print(f"Cannot parse response '{ret_solution}' to double value.")
            else:
                print(f"Request failed with status code {response.status_code}")
        except requests.RequestException as exc:
            print(f"An error occurred while requesting {exc.request.url!r}.")

        return exact_solution
